imap_processing/utils.py

```diff
@@ -299,10 +299,18 @@ def _create_minimum_dtype_array(values: list, dtype: str) -> npt.NDArray:
     array : np.array
         The array of values.
     """
+    # Create an initial array and then try to safely cast it to the desired dtype
+    x = np.asarray(values)
     try:
-        return np.array(values, dtype=dtype)
+        # ValueError: when trying to cast strings (enum states) to ints
+        y = x.astype(dtype, copy=False)
+        # We need to compare the arrays to see if we trimmed any values by
+        # casting to a smaller datatype (e.g. float64 to uint8, 2.1 to 2)
+        if np.array_equal(x, y):
```

Comment:
This is all somewhat unfortunate in how many "arrays" we'll be creating here. If anyone has ideas for a better way to do this I'm interested in hearing it.

We are potentially creating 3 separate arrays:
1. initial asarray from a list -> creates an int64
2. Trying to cast to a smaller dtype -> creates a uint8
3. Doing this array comparison in a higher-order dtype space.

They are all happening in numpy c-level functions so shouldn't be a huge hit, but it would be nice to not do this array comparison just to check loss of precision.

Reply:
My thought would be to modify the `_get_minimum_numpy_datatype` function to take the derived value logic into account and return the appropriate datatype.

Reply:
Really good suggestion, thanks.

I opened up: https://github.com/IMAP-Science-Operations-Center/imap_processing/pull/786
with an alternative approach, which is actually able to get rid of the numpy try/except array creation too. The one downside is there isn't a great way to test all of the XTCE stuff currently :-/ I did test it on the new SWE XTCE file that was failing and it looks good, but I don't have a great way to parameterize that PR like we do with this one.